src/player.c

```diff
@@ -1926,3 +1926,120 @@ void xmp_get_frame_info(xmp_context opaque, struct xmp_frame_info *info)
 		}
 	}
 }
+
+int xmp_add_channels(xmp_context opaque, int start, int num)
+{
+	struct context_data *ctx = (struct context_data *)opaque;
+	struct module_data *m = &ctx->m;
+	struct player_data *p = &ctx->p;
+	struct xmp_module *mod = &m->mod;
+	int pat, c, t;
+
+	if (num == 0)
+		return 0;
+	if (num > 0) {
+		if (mod->chn + num > XMP_MAX_CHANNELS)
+			return -XMP_ERROR_INVALID;
+	}
+	else {
+		if (mod->chn + num < 0)
+			return -XMP_ERROR_INVALID;
+	}
+
+	if (start < 0)
+		start = mod->chn + 1 + start;
+	if (start < 0 || start > mod->chn)
+		return -XMP_ERROR_INVALID;
+
+	if (num < 0) {
+		if (start - num > mod->chn)
+			return -XMP_ERROR_INVALID;
```

Comment:
Should `start + num < 0` be checked here? That would ensure that a number of channels larger than the position they are being removed from can not be removed. I don't know that would actually hurt anything though.

I'm not sure a check for `start - num > mod->chn` is necessary since the following loops are being bounded against `mod->chn + num`. But, for example, if you're removing 8 tracks starting from position 10 of a 12 track module, this check would result in 18 > 12 and fail (when that operation should be completely valid).